Add Thycotic DevOps Secrets Vault support

amigus opened the pull request:
##### SUMMARY
Add a Credential Plugin for Thycotic DevOps Secrets Vault 

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - API
 - UI

##### AWX VERSION
```
awx: 19.1.0
```

amigus committed:
Add Thycotic DevOps Secrets Vault support

amigus committed:
Add Thycotic DevOps Secrets Vault support

amigus committed:
Backout accidential changes to settings/defaults

amigus committed:
Backout accidential changes to settings/defaults

amigus committed:
Fix test_credential.py
- Add 'thycotic_dsv' to the list of expected credential types

amigus committed:
Add license for python-dsv-sdk to docs/licenses

shanemcd commented on awx/main/credential_plugins/dsv.py:
```diff
@@ -0,0 +1,71 @@
+dsv_plugin = CredentialPlugin(
+    'Thycotic DevOps Secrets Vault',
+    dsv_inputs,
+    lambda **kwargs: eval(
```
Do we really need to use `eval` here?

amigus replied:
Yes and no. There has to be some mechanism to allow users to "select" the field or fields they want from the secret data. Eval was the simplest. At one point, I had a TODO to constrain it with a safety check. Now that you've raised the issue, I'll look at that and make some changes.

The alternative is to allow for the use of "JSONpath," which is the pseudo-equivalent of XPath, but I opted against that because I don't see much adoption and/or a prevalent library that does it.

shanemcd replied:
@amigus Any luck with that? This will definitely need to be addressed before we can consider this PR.

amigus replied:
Understood and no disagreement. I'll resolve it ASAP as there are people waiting on the plugin.

matburt commented on awx/main/credential_plugins/dsv.py:
```diff
@@ -0,0 +1,71 @@
+        }
+    )
+
+if to_locale(get_language()) != 'en_US':
```
Can you explain what's going on here?

amigus replied:
Why, a dubious assumption, of course! 😁

The TLD reflects where the cloud instance lives and it defaults to '.com' which is a US deployment.

The assumption is that the user will need to use a non-default TLD for DSV only when AWX is running in a non-US region.

I originally had an override in settings.py but I removed it.

matburt replied:
It just seems strange, does it not make more sense to just have them enter the base url on its own? If the tld or, god forbid, the url structure needs to change then it would require a code change that wouldn't be picked up until the next Tower release. Seems like something that could get broken pretty easily?

amigus commented:
Customers want the plugin ASAP so I'm going to remove the conditionals and swap the eval for a simple match. Give me a day or two.

amigus commented:
Welp, I tried to go for the gusto and implement expression evaluation using pyjq but I couldn't get the `make docker-compose-build` to succeed once I added pyjq to requirements.in and `./updater.sh`. I got:

> #24 68.93 ERROR: Could not build wheels for pyjq which use PEP 517 and cannot be installed directly

If anyone has a fix for that LMK otherwise I'm stuck implementing the minimum. 😒

amigus committed:
Refactor to simplify
- Remove eval and the expression metadata attribute
- Remove locale conditional on the tld field

amigus commented:
This version no longer has an `eval` nor does it check the locale. Reasonable?

amigus committed:
Remove unused imports to fix build.

matburt commented:
I think you might have messed up the rebase on this

![image](https://user-images.githubusercontent.com/89544/121041345-ab620080-c780-11eb-9dc2-6f0c27e1c6a7.png)

amigus commented:
Argh! So just rebase to devel and push?

amigus commented:
> Argh! So just rebase to devel and push?

Pardon my ignorance but do I just need to rebase the branch onto devel and push it?

beeankha commented:
Hey @amigus , hopefully the following will help:

1) Run `git remote -v` and note the name of the remote that's the upstream AWX repo
2) While on your working branch (the one for this PR), rebase with `devel` from the name of that upstream remote via `git pull --rebase [name_of_upstream_remote] devel`

amigus commented:
Did I do it right? Are we ready to merge?

matburt reviewed (approved):
This all looks reasonable to me

shanemcd reviewed (approved):
Thanks for seeing this through!

amigus commented:
Thanks a bunch, fellas!  FYI, I'll be submitting a 'strikingly similar' one for Thycotic Secret Server, shortly.
